libraries/chain/hotstuff/hs_pseudo

```diff
@@ -0,0 +1,345 @@
+	    return result;
+	}
+}
+
+struct block_header_state {
+	//existing block_header_state members
+	sha256						id; //b_leaf under hotstuff
+	[...] //other existing block_header_state members
+	//new additions
+	sha256 						finalizer_digest;
+	block_header_state_core 	core;
+	incremental_block_mtree		proposal_mtree;
+	incremental_block_mtree		finality_mtree;
+	uint32_t					policy_generation;
+}
+
+struct block_handle {
+	block_header_state_ptr 						bhs;
+	finalizer_policy_ptr 						fp;	
+	std::optional<pending_quorum_certificate>	pending_qc; 
+	std::optional<valid_quorum_certificate>		valid_qc; 
+	uint64_t id() const {return bhs->id;}
+	uint64_t get_height() const {return block_header::num_from_id(bhs->id);}
+	quorum_certificate get_best_qc() { [...] //return the best QC available }
+}
+
+//this structure holds the required information and methods for the Hotstuff algorithm. It is derived from a block and block_header content, notably extensions 
+struct hs_proposal {
+	//may not exist in final implementation, subject to change 
+	block_id_type 						block_id; //computed, to be replaced with proposal_digest eventually
+	uint32_t 							get_height(); //from block_id
+	block_timestamp_type 				timestamp; //from block header
+	//qc specific information
+	uint32_t							last_qc_block_height; //from block header extension
+	bool 								is_last_qc_strong; //from block header extension
+	valid_quorum_certificate		 	qc; //from block extension
+};
+
+struct valid_quorum_certificate {
+	hs_bitset 				strong_bitset;
+	optional<hs_bitset> 	weak_bitset; //omitted if strong qc
+	bls_signature 			signature; //set to strong_signature if strong qc, set to strong_signature + weak_signature if weak qc
+	
+	//constructor used for strong qc
+	valid_quorum_certificate(hs_bitset b, bls_signature s) : 
+	strong_bitset(b),
+	signature(s){}
+
+	//constructor used for weak qc
+	valid_quorum_certificate(hs_bitset sb, hs_bitset wb,  bls_signature s) :
+	strong_bitset(sb),
+	weak_bitset(wb),
+	signature(s){}
+
+	bool is_strong() {if (weak_bitset.has_value()) return false; else return true; }
+}
+
+struct pending_quorum_certificate {
+	hs_bitset 		strong_bitset;
+	bls_signature 	strong_signature;
+	hs_bitset 		weak_bitset;
+	bls_signature 	weak_signature;
+	bool strong_quorum_met() [...] //abstracted, returns true if a strong quorum is met, false otherwise
+	bool weak_quorum_met()[...] //abstracted, returns true if a weak quorum is met, false otherwise
+}
+
+struct quorum_certificate {
+	uint32_t block_height;
+	valid_quorum_certificate qc;
+}
+
+struct hs_vote_message {
+	block_id_type   block_id; //temporary, probably not needed later
+	sha256          proposal_digest; //proposal digest
+	bls_public_key  finalizer_key;
+	bls_signature   sig;
+	bool 			weak; //indicate if vote is weak, strong otherwise
+};
+
+
+//added as a block_header extension before signing
+struct hotstuff_header_extension {
+	uint32_t					last_qc_block_height;
```

Comment:
Not sure why this is only the height. Doesn't it need to be block id or block digest?

Reply:
I think for the header extension we only store the last_qc_block_height and is_last_qc_strong.

Assuming the block is valid, we should always be able to locate an ancestor block at last_qc_block_height on that branch and get its id. 

See : 
https://github.com/AntelopeIO/leap/issues/1911